File: shared-backend/crmApp/serializers/message.py
"""
Message Serializers
"""

from rest_framework import serializers
from crmApp.models import Message, Conversation
from crmApp.serializers.auth import UserSerializer


class MessageSerializer(serializers.ModelSerializer):
    """Serializer for Message model"""
    
    sender = UserSerializer(read_only=True)
    recipient = UserSerializer(read_only=True)
    sender_id = serializers.IntegerField(write_only=True, required=False)
    recipient_id = serializers.IntegerField(write_only=True)
    sender_email = serializers.CharField(source='sender.email', read_only=True)
    recipient_email = serializers.CharField(source='recipient.email', read_only=True)
    
    class Meta:
        model = Message
        fields = [
            'id', 'sender', 'recipient', 'sender_id', 'recipient_id',
            'sender_email', 'recipient_email', 'message_type', 'subject',
            'content', 'organization', 'is_read', 'read_at',
            'related_lead', 'related_deal', 'related_customer',
            'attachments', 'created_at', 'updated_at'
        ]
        read_only_fields = ['id', 'created_at', 'updated_at', 'is_read', 'read_at']
    
    def create(self, validated_data):
        """Create message with sender from request"""
        request = self.context.get('request')
        if request and request.user:
            validated_data['sender'] = request.user
        
        # Get recipient from recipient_id
        recipient_id = validated_data.pop('recipient_id', None)
        if recipient_id:
            from django.contrib.auth import get_user_model
            User = get_user_model()
            validated_data['recipient'] = User.objects.get(id=recipient_id)
        
        return super().create(validated_data)


class MessageCreateSerializer(serializers.Serializer):
    """Serializer for creating messages"""
    
    recipient_id = serializers.IntegerField(required=True)
    content = serializers.CharField(required=True, max_length=5000)
    subject = serializers.CharField(required=False, allow_blank=True, max_length=255)
    related_lead_id = serializers.IntegerField(required=False, allow_null=True)
    related_deal_id = serializers.IntegerField(required=False, allow_null=True)
    related_customer_id = serializers.IntegerField(required=False, allow_null=True)
    attachments = serializers.ListField(
        child=serializers.DictField(),
        required=False,
        allow_empty=True
    )
    
    def validate_recipient_id(self, value):
        """Validate recipient exists"""
        from django.contrib.auth import get_user_model
        User = get_user_model()
        try:
            User.objects.get(id=value)
        except User.DoesNotExist:
            raise serializers.ValidationError("Recipient not found")
        return value


class ConversationSerializer(serializers.ModelSerializer):
    """Serializer for Conversation model"""
    
    participant1 = UserSerializer(read_only=True)
    participant2 = UserSerializer(read_only=True)
    last_message = MessageSerializer(read_only=True)
    other_participant = serializers.SerializerMethodField()
    unread_count = serializers.SerializerMethodField()
    
    class Meta:
        model = Conversation
        fields = [
            'id', 'participant1', 'participant2', 'other_participant',
            'organization', 'last_message', 'last_message_at',
            'unread_count_participant1', 'unread_count_participant2',
            'unread_count', 'created_at', 'updated_at'
        ]
        read_only_fields = ['id', 'created_at', 'updated_at']
    
    def get_other_participant(self, obj):
        """Get the other participant (not the current user)"""
        request = self.context.get('request')
        if request and request.user:
            if obj.participant1 == request.user:
                return UserSerializer(obj.participant2).data
            return UserSerializer(obj.participant1).data
        return None
    
    def get_unread_count(self, obj):
        """Get unread count for current user"""
        request = self.context.get('request')
        if request and request.user:
            return obj.get_unread_count(request.user)
        return 0

File: app-frontend/app/src/main/java/too/good/crm/data/model/Message.kt
package too.good.crm.data.model

import com.google.gson.annotations.SerializedName

/**
 * Message data model matching backend API
 * Endpoint: /api/messages/
 * Simple user-to-user messaging model
 */
data class Message(
    val id: Int,
    val sender: MessageUser,
    val recipient: MessageUser,
    val content: String,
    val subject: String?,
    @SerializedName("message_type") val messageType: String = "text",
    @SerializedName("is_read") val isRead: Boolean,
    @SerializedName("read_at") val readAt: String?,
    val organization: Int?,
    @SerializedName("related_lead") val relatedLead: Int?,
    @SerializedName("related_deal") val relatedDeal: Int?,
    @SerializedName("related_customer") val relatedCustomer: Int?,
    val attachments: List<Any>? = null,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String
)

/**
 * User information in message
 */
data class MessageUser(
    val id: Int,
    val email: String,
    @SerializedName("first_name") val firstName: String?,
    @SerializedName("last_name") val lastName: String?
) {
    fun getDisplayName(): String {
        val name = listOfNotNull(firstName, lastName).joinToString(" ").trim()
        return name.ifBlank { email.substringBefore("@") }
    }
}

/**
 * Conversation data model (simplified for direct messaging)
 */
data class Conversation(
    val id: Int,
    val participant1: MessageUser,
    val participant2: MessageUser,
    @SerializedName("other_participant") val otherParticipant: MessageUser,
    @SerializedName("last_message") val lastMessage: Message?,
    @SerializedName("last_message_at") val lastMessageAt: String?,
    @SerializedName("unread_count") val unreadCount: Int,
    val organization: Int?
)

/**
 * Request for sending a message
 */
data class SendMessageRequest(
    @SerializedName("recipient_id") val recipientId: Int,
    val content: String,
    val subject: String? = null,
    @SerializedName("related_lead_id") val relatedLeadId: Int? = null,
    @SerializedName("related_deal_id") val relatedDealId: Int? = null,
    @SerializedName("related_customer_id") val relatedCustomerId: Int? = null,
    val attachments: List<Any>? = null
)

/**
 * Response for unread count
 */
data class UnreadCountResponse(
    @SerializedName("unread_count") val unreadCount: Int
)

/**
 * Response wrapping conversations list
 */
data class ConversationsListResponse(
    val count: Int,
    val next: String?,
    val previous: String?,
    val results: List<Conversation>
)

